src/main/java/gov/cabinetoffice/gap/applybackend/service/GrantMandatoryQuestionService.java

```diff
@@ -96,13 +89,39 @@ public GrantMandatoryQuestions createMandatoryQuestion(GrantScheme scheme, Grant
     }
 
 
-    public GrantMandatoryQuestions updateMandatoryQuestion(GrantMandatoryQuestions grantMandatoryQuestions) {
+    public GrantMandatoryQuestions updateMandatoryQuestion(GrantMandatoryQuestions grantMandatoryQuestions, GrantApplicant grantApplicant) {
+        if(grantMandatoryQuestions.getStatus().equals(GrantMandatoryQuestionStatus.COMPLETED)) {
+            final Submission submission = grantMandatoryQuestions.getSubmission();
+            grantMandatoryQuestions.setGapId(submission == null ? generateGapId(grantApplicant.getId()) : submission.getGapId());
+        }
         return grantMandatoryQuestionRepository
                 .findById(grantMandatoryQuestions.getId()) //TODO there is no need for the additional database call here
                 .map(mandatoryQuestion -> grantMandatoryQuestionRepository.save(grantMandatoryQuestions))
                 .orElseThrow(() -> new NotFoundException(String.format("No Mandatory Question with id %s was found", grantMandatoryQuestions.getId())));
     }
 
+    private String generateGapId(final Long userId) {
```

Comment:
If we need to reuse this across different classes can we shift it into a static utility class instead of duplicating the code please? If you need a hand I can help out with that.